pyfolio/perf_attrib.py

```diff
@@ -0,0 +1,134 @@
+#
+# Copyright 2017 Quantopian, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import division
+import pandas as pd
+
+import empyrical
+from pyfolio.pos import get_percent_alloc
+from pyfolio.utils import print_table
+
+
+def perf_attrib(factor_loadings, positions, factor_returns, returns):
+    """
+    Does performance attribution given risk info.
+
+    Parameters
+    ----------
+    factor_loadings : pd.DataFrame
+        Factor loadings for all days in the date range, with date and ticker as
+        index, and factors as columns.
+        - Example:
+                               momentum  reversal
+            dt         ticker
+            2017-01-01 AAPL   -1.592914  0.852830
+                       TLT     0.184864  0.895534
+                       XOM     0.993160  1.149353
+            2017-01-02 AAPL   -0.140009 -0.524952
+                       TLT    -1.066978  0.185435
+                       XOM    -1.798401  0.761549
+
+    positions: pd.DataFrame
+        Daily holdings (in dollars or percentages), indexed by date.
+        Will be converted to percentages if positions are in dollars.
+        - Example:
+                        AAPL  TLT  XOM
+            2017-01-01    71   93   10
+            2017-01-02    71   16   71
+            2017-01-03    46   43   63
+
+    factor_returns : pd.DataFrame
+        Returns by factor, with date as index and factors as columns
+        - Example:
+                        momentum  reversal
+            2017-01-01  0.002779 -0.005453
+            2017-01-02  0.001096  0.010290
+
+    returns : pd.Series
+        Returns for each day in the date range.
+        - Example:
+            2017-01-01   -0.017098
+            2017-01-02    0.002683
+            2017-01-03   -0.008669
+
+    Returns
+    -------
+
+    perf_attribution : pd.DataFrame
+        df with factors, common returns, and specific returns as columns,
+        and datetimes as index
+        - Example:
+                        momentum  reversal  common_returns  specific_returns
```

Comment:
Actually, you're right, we also need to return the individual perf-attributed factor-returns from `perf_attrib_by_factor`. You would need to add that below in the `pd.DataFrame` creation.